src/middlewared/middlewared/plugins/pool_/encrypt_disks.py

```diff
@@ -0,0 +1,61 @@
+import os
+from tempfile import NamedTemporaryFile
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+from middlewared.service import private, Service
+
+
+class PoolService(Service):
+
+    @private
+    def encrypt_disks(self, job, disks, options):
+        """
+        This GELI encrypts all the disks in `disks`.
+        NOTE: this is only called on pool.update since GELI was deprecated
+        for all newly created zpools.
+        """
+        # we do not allow the creation of new GELI based zpools, however,
+        # we still have to support users who created them before we deprecated
+        # the parameters in the API. This is here for those users.
+        total_disks = len(disks)
+        pass_file = None
+        if options.get('passphrase'):
+            # called when attaching disks to a GELI encrypted zpool
+            with NamedTemporaryFile(mode='w+', dir='/tmp', delete=False) as p:
+                os.chmod(p.name, 0o600)
+                p.write(options['passphrase'])
+                p.flush()
+                options['passphrase_path'] = pass_file = p.name
+
+        with ThreadPoolExecutor(max_workers=16) as exc:
+            # The called methods `subprocess` out many times for each disk
+            # so this is painful on systems with large amount of disks.
+            # This is, unfortunately, the best we can really do right now
+            # without implementing gpart code natively in python.
+            # (cython or ctypes wrapper maybe???)
+            futures = [
+                exc.submit(
+                    self.middleware.call_sync(
```

Comment:
We should not be submitting to the executor this way - we should be passing it a callable not calling the callable and then passing it
```
def dummy():
    return None
with ThreadPoolExecutor(max_workers=1) as exc:
    futures = [exc.submit(dummy) for i in range(5)]
    results = []
    for idx, fut in enumerate(as_completed(futures), start=1):
        results.append(fut.result())
print(results)

```

Reply:
Also if i am not wrong, this means that we are currently serializing the calls and not actually doing it in parallel